### PR TITLE
Refactor Act patching to use typed path-based patches instead of delta objects

### PR DESCRIPTION
**Pull request by toyamarinyon**
## Summary
Refactored the Act patching system to use a more type-safe and structured approach with explicit patch objects instead of delta objects.

## Changes
- Replaced the `delta` object with an array of explicit `patches` in the Act patching API
- Created a new type-safe patch creator system with strongly typed paths and operations
- Added comprehensive test coverage for the new patching system
- Enhanced the Act model with additional fields for tracking durations, usage metrics, and annotations
- Updated the workflow designer UI to use the new patching API
- Improved error handling and sequence management in the Act execution flow

## Testing
- Added extensive unit tests for the patch creators and patch application logic
- Tests cover various scenarios including edge cases, immutability, and type safety

## Other Information
This change improves type safety and makes the patching API more explicit and maintainable. The new approach allows for better static analysis and reduces the chance of runtime errors when modifying Act objects.

### CODE REVIEW

**Comment by toyamarinyon** [comment]
* **#1538** <a href="https://app.graphite.dev/github/pr/giselles-ai/giselle/1538?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/giselles-ai/giselle/1538?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/giselle/src/engine/acts/object/patch-object.ts [hunk 1]
```diff
@@ -1,100 +1,55 @@
...
-				target[lastKey] = patch.set;
-			} else if ("increment" in patch) {
-				const inc = patch.increment ?? 0;
-				target[lastKey] = current + inc;
-			} else if ("decrement" in patch) {
-				const dec = patch.decrement ?? 0;
-				target[lastKey] = current - dec;
+		// Apply the patch
+		if ("set" in patch) {
+			target[lastKey] = patch.set;
```
[the remark this answers is not in the document]
Comment (toyamarinyon) [reply packages/giselle/src/engine/acts/object/patch-object.ts:56]: 1. Added validation to check for dangerous keys (__proto__, constructor, prototype) in the path
  2. Throws an error if any part of the path contains these dangerous keys, preventing prototype pollution
  3. Added comprehensive tests to ensure:
    - Prototype pollution attempts are blocked
    - Legitimate paths still work correctly
    - Edge cases are handled properly

[the remark this answers is not in the document]
Comment (toyamarinyon) [reply packages/giselle/src/engine/acts/object/patch-object.ts:56]: Yes, I changed the code in a23c5deafb28863e91b5408998f6f85061ebef1fd

**Review by shige — APPROVED** [review]
Thank you! ♻️

**Comment by toyamarinyon** [comment]
### Merge activity

* **Jul 24, 3:07 AM UTC**: @toyamarinyon merged this pull request with [Graphite](https://app.graphite.dev/github/pr/giselles-ai/giselle/1538).

File: packages/giselle/src/engine/acts/patch-act.ts [hunk 2]
```diff
@@ -1,33 +1,30 @@
...
+		{ path: "updatedAt", set: Date.now() },
+	];
 
-	// Apply the patch
-	const updatedAct = patchActObject(currentAct, {
-		...args.delta,
-		updatedAt: { set: Date.now() },
-	});
+	// Apply the patches
+	const updatedAct = patchActObject(currentAct, ...allPatches);
```
[the remark this answers is not in the document]
Comment (toyamarinyon) [reply packages/giselle/src/engine/acts/patch-act.ts:27]: I'll fix on https://github.com/giselles-ai/giselle/pull/1539

File: packages/giselle/src/engine/acts/start-act.ts [hunk 3]
```diff
@@ -167,63 +248,59 @@ export async function startAct(
...
 		context: args.context,
 		actId: args.actId,
-		delta: {
-			status: { set: "completed" },
-			"duration.wallClock": { set: Date.now() - flowStart },
-		},
+		patches: [
+			patches.status.set("completed"),
+			patches.duration.wallClock.set(Date.now() - flowStart),
+		],
```
[the remark this answers is not in the document]
Comment (toyamarinyon) [reply packages/giselle/src/engine/acts/start-act.ts:304]: I'll fix on https://github.com/giselles-ai/giselle/pull/1540

File: packages/giselle/src/engine/acts/start-act.ts [hunk 4]
```diff
@@ -27,93 +28,96 @@ export interface StartActCallbacks {
...
+		await args.onError(e, {
+			...args.step,
+			duration: Date.now() - startedAt,
+		});
+		return;
 	}
+	await args.onFinish({
+		...args.step,
+		duration: Date.now() - startedAt,
+	});
```
[the remark this answers is not in the document]
Comment (toyamarinyon) [reply packages/giselle/src/engine/acts/start-act.ts:103]: I'll fix on https://github.com/giselles-ai/giselle/pull/1541
